src/peft/utils/other.py

```diff
@@ -1476,16 +1543,36 @@ def set_additional_trainable_modules(model, peft_config, model_config, adapter_n
                 activate_adapter=activate_adapter,
             )
 
-        # There might be the possibility that we have output weights that are tied to the input weights.
-        # In that case we will tie any module that wants tied weights to the token adapter to make sure that
-        # any modification is reflected in the tied layers as well.
-        if (
-            model_config.get("tie_word_embeddings", False)
-            # some models may be misconfigured to have weight tying enabled but don't define tied weights keys
-            and model._tied_weights_keys is not None
+        # Case 1: weights NOT tied + ensure_weight_tying=True -> WARNING
+        if not weights_tied and ensure_weight_tying:
+            warnings.warn(
+                "ensure_weight_tying=True but the model does not have tied weights "
+                "(tie_word_embeddings=False). Weight tying will not be applied for trainable_token_indices."
+            )
+
+        # Case 2: weights tied + ensure_weight_tying=True + different indices -> ERROR
+        if weights_tied and ensure_weight_tying and has_both_layers and indices_mismatch:
+            # Build more generic error message showing the conflicting layers
+            tied_layers_info = ", ".join([f"{key}: {target_layers[key]}" for key in tied_layer_keys])
+            raise ValueError(
```

Comment:
Can we not raise this error immediately after `indices_mismatch` was determined? The earlier we can raise, the better. It should also make the check simpler, as we only need to check for `if indices_mismatch`.